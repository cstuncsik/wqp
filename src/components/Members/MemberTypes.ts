import { Dictionary } from 'helpers/types'

export interface Member {
  person_id: string
  firstname: string
  lastname: string
  title: string
  business_unit: string
  is_user: boolean
}

export interface MembersFCProps {
  members: Member[]
}

export interface MemberAccessLevel {
  id: 'read' | 'write' | 'admin'
  label: string
}

export interface MemberRoleSettings {
  role: 'customer' | 'employee' | 'manager'
  access_level: MemberAccessLevel['id']
}

export type MemberWithRole = Member & MemberRoleSettings

export interface MemberRole {
  id: MemberRoleSettings['role']
  label: string
  accessLevels: MemberAccessLevel['id'][]
}

export interface MemberRowFCProps {
  idx: number
  selectedMember: MemberWithRole
  selectableMembers: MemberWithRole[]
  removeMember: (idx: number) => void
  updateMember: (idx: number, data: MemberWithRole) => void
  switchMember: (idx: number, newMemberId: string) => void
  roles: Dictionary<MemberRole>
  accessLevels: Dictionary<MemberAccessLevel>
}

export type MemberResult = Partial<MemberWithRole>
