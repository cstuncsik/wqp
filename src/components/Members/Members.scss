.Members {
  display: grid;
  padding: 20px;
  grid-template-rows: 1fr auto auto;
  grid-row-gap: 20px;
  grid-area: Members;
}

.Members-data {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  grid-auto-rows: min-content;
  grid-column-gap: 20px;
  grid-row-gap: 20px;
  align-items: center;
  border-top: 2px solid #ccc;
}

.Members-result {
  padding: 20px 0 0;
  border-top: 2px solid #ccc;

  pre {
    margin: 0;
    padding-top: 20px;
    color: #fff;
  }
}
