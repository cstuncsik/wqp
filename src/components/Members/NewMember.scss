.New-member {
  display: flex;
  padding: 20px;
  align-items: center;
  justify-content: flex-end;
  grid-area: New-member;
}

.New-member-overlay {
  position: absolute;
  display: flex;
  align-items: center;
  justify-content: center;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background: rgba(0, 0, 0, 0.75);
}

.New-member-content {
  background-color: #282c34;
  border: 2px solid #ccc;
  padding: 20px;
}

.New-member-content-header {
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.New-member-content-footer {
  display: flex;
  align-items: center;
  justify-content: flex-end;
}

.New-member-content-form {
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 20px;
}

.New-member-content-form-input {
  input[type='text'] {
    display: block;
    width: 100%;
  }

  small {
    display: block;
  }
}
