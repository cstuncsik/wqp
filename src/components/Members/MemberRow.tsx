import React, { FC, Fragment } from 'react'

import { MemberAccessLevel, MemberRole, MemberRoleSettings, MemberRowFCProps } from 'components/Members/MemberTypes'

export const MemberRow: FC<MemberRowFCProps> = ({
  idx,
  selectedMember,
  selectableMembers,
  removeMember,
  updateMember,
  switchMember,
  roles,
  accessLevels
}) => {
  return (
    <Fragment>
      <select
        name="member"
        disabled={selectableMembers.length === 1}
        value={selectedMember.person_id}
        onChange={({ target }) => {
          switchMember(idx, target.value)
        }}
      >
        {selectableMembers.map(option => (
          <option key={option.person_id} value={option.person_id}>
            {option.firstname} {option.lastname}
          </option>
        ))}
      </select>
      <select
        name="role"
        value={selectedMember.role}
        onChange={({ target }) => {
          selectedMember.role = target.value as MemberRoleSettings['role']
          selectedMember.access_level = roles[selectedMember.role].accessLevels[0]
          updateMember(idx, selectedMember)
        }}
      >
        {Object.keys(roles).map(key => {
          const role: MemberRole = roles[key]
          return (
            <option key={role.id} value={role.id}>
              {role.label}
            </option>
          )
        })}
      </select>
      <select
        name="accessLevel"
        disabled={roles[selectedMember.role].accessLevels.length === 1}
        value={selectedMember.access_level}
        onChange={({ target }) => {
          selectedMember.access_level = target.value as MemberAccessLevel['id']
          updateMember(idx, selectedMember)
        }}
      >
        {roles[selectedMember.role].accessLevels.map(accessLevel => (
          <option key={accessLevel} value={accessLevels[accessLevel].id}>
            {accessLevels[accessLevel].label}
          </option>
        ))}
      </select>
      <div>
        <button type="button" onClick={() => removeMember(idx)}>
          X
        </button>
      </div>
    </Fragment>
  )
}
