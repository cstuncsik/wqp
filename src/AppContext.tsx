import React, { FC, Fragment, createContext, useState } from 'react'

import { AppContextProps } from 'AppTypes'
import { MessageItem } from 'components/Message/MessageItem'
import { Message } from 'components/Message/MessageTypes'
import { Member } from 'components/Members/MemberTypes'
import { uuidv4 } from 'helpers/string'

const defaultAppContextValue: AppContextProps = {
  messages: [],
  addMessage: message => {},
  availableMembers: [],
  setAvailableMembers: availableMembers => {}
}

const AppContext = createContext(defaultAppContextValue)
const { Provider } = AppContext

const AppProvider: FC = ({ children }) => {
  const [messages, setMessages] = useState<Message[]>([])
  const [availableMembers, setAvailableMembers] = useState<Member[]>([])

  const addMessage = (message: Omit<Message, 'id'>): void => {
    setMessages([...messages, { ...message, id: uuidv4() }])
  }

  const removeMessage = (message: Message): void => {
    setMessages(messages.filter(msg => msg.id !== message.id))
  }
  return (
    <Provider
      value={{
        messages,
        addMessage,
        availableMembers,
        setAvailableMembers
      }}
    >
      <Fragment>
        {messages.length > 0 && (
          <ul className="App-messages">
            {messages.map(message => (
              <MessageItem
                key={message.id}
                {...{
                  message,
                  onClose: () => {
                    removeMessage(message)
                  }
                }}
              />
            ))}
          </ul>
        )}
        {children}
      </Fragment>
    </Provider>
  )
}

export { AppContext, AppProvider }
