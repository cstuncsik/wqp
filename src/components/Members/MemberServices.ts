import { AxiosPromise } from 'axios'
import { http } from 'services/http'
import { Member } from 'components/Members/MemberTypes'

export const getMe = (): AxiosPromise<Member> => http.get('/me')
export const getMembers = (): AxiosPromise<Member[]> => http.get('/members')
