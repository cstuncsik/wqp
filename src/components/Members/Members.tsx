import React, { FC, useEffect, useState, Fragment } from 'react'

import {
  Member,
  MemberAccessLevel,
  MemberRole,
  MemberRoleSettings,
  MembersFCProps,
  MemberWithRole,
  MemberResult
} from 'components/Members/MemberTypes'
import { Dictionary } from 'helpers/types'
import { MemberRow } from 'components/Members/MemberRow'
import { pick } from 'helpers/object'

import 'components/Members/Members.scss'

const roles: Dictionary<MemberRole> = {
  customer: {
    id: 'customer',
    label: 'Customer',
    accessLevels: ['read', 'write']
  },
  employee: {
    id: 'employee',
    label: 'Employee',
    accessLevels: ['write', 'admin']
  },
  manager: {
    id: 'manager',
    label: 'Manager',
    accessLevels: ['admin']
  }
}

const accessLevels: Dictionary<MemberAccessLevel> = {
  read: {
    id: 'read',
    label: 'Read'
  },
  write: {
    id: 'write',
    label: 'Write'
  },
  admin: {
    id: 'admin',
    label: 'Admin'
  }
}

const memberDefaultRoleProps: MemberRoleSettings = { role: 'customer', access_level: 'read' }

export const Members: FC<MembersFCProps> = ({ members }) => {
  const [selectedMembers, setSelectedMembers] = useState<MemberWithRole[]>([])
  const [availableMembers, setAvailableMembers] = useState<Member[]>([])
  const [membersResult, setMembersResult] = useState<{ members: MemberResult[] }>({ members: [] })

  const addMember = (): void => {
    const added: MemberWithRole = { ...availableMembers.shift()!, ...memberDefaultRoleProps }
    setSelectedMembers(selectedMembers.concat(added))
  }

  const removeMember = (idx: number): void => {
    const removed: MemberWithRole = selectedMembers.splice(idx, 1)[0]
    setSelectedMembers(selectedMembers)
    setAvailableMembers(availableMembers.concat(removed))
  }

  const updateMember = (idx: number, data: MemberWithRole): void => {
    selectedMembers[idx] = data
    setSelectedMembers([...selectedMembers])
  }

  const switchMember = (idx: number, newMemberId: string): void => {
    const newMemberIdx: number = availableMembers.findIndex(member => member.person_id === newMemberId)
    const removedAvailableMember: Member = availableMembers.splice(newMemberIdx, 1)[0]
    const added: MemberWithRole = { ...removedAvailableMember, ...memberDefaultRoleProps }
    const removed: MemberWithRole = selectedMembers.splice(idx, 1, added)[0]
    setAvailableMembers(availableMembers.concat(removed))
  }

  const createMembersResult = (): void => {
    setMembersResult({ members: selectedMembers.map(pick(['person_id', 'role', 'access_level'])) })
  }

  useEffect(() => {
    setAvailableMembers([...members])
  }, [members])

  return (
    <div className="Members" data-testid="members">
      {members.length > 0 ? (
        <Fragment>
          <div className="Members-data">
            <h2>Member</h2>
            <h2>Role</h2>
            <h2>Access Level</h2>
            <span>
              <button type="button" onClick={addMember} disabled={!availableMembers.length}>
                Add new member
              </button>
            </span>
            {selectedMembers.map((selectedMember, idx) => {
              const selectableMembers: MemberWithRole[] = availableMembers.reduce(
                (additionalSelectableMembers, member) =>
                  selectedMember.person_id !== member.person_id
                    ? additionalSelectableMembers.concat({ ...member, role: 'customer', access_level: 'read' })
                    : additionalSelectableMembers,
                [selectedMember]
              )
              return (
                <MemberRow
                  key={selectedMember.person_id}
                  {...{
                    idx,
                    selectedMember,
                    selectableMembers,
                    removeMember,
                    updateMember,
                    switchMember,
                    roles,
                    accessLevels
                  }}
                />
              )
            })}
          </div>
          <div className="Members-result">
            <button type="button" onClick={createMembersResult}>
              Save result
            </button>
            {membersResult.members.length > 0 && <pre>{JSON.stringify(membersResult, null, 2)}</pre>}
          </div>
        </Fragment>
      ) : (
        <h2>No available members</h2>
      )}
    </div>
  )
}
