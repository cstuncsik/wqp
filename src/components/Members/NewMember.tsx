import React, { FC, useState, ChangeEvent, useContext, useCallback } from 'react'

import { Member } from 'components/Members/MemberTypes'
import 'components/Members/NewMember.scss'
import { uuidv4 } from 'helpers/string'
import { AppContext } from 'AppContext'

const createEmptyMember = (): Member => ({
  firstname: '',
  lastname: '',
  title: '',
  person_id: '',
  business_unit: '',
  is_user: false
})

export const NewMember: FC = () => {
  const [opened, setOpened] = useState<boolean>(false)
  const [newMember, setNewMember] = useState<Member>(createEmptyMember)
  const { availableMembers, setAvailableMembers, addMessage } = useContext(AppContext)

  const openOverlay = (): void => {
    setOpened(true)
    setNewMember(createEmptyMember())
  }

  const closeOverlay = (): void => {
    setOpened(false)
  }

  const updateMemberPropFromString = (e: ChangeEvent<HTMLInputElement>) => {
    setNewMember({ ...newMember, [e.target.name]: e.target.value })
  }

  const createNewMember = () => {
    setAvailableMembers([...availableMembers, { ...newMember, person_id: uuidv4() }])
    closeOverlay()
    addMessage({ type: 'success', text: 'Succesfully created a new member' })
  }

  const firstnameInput = useCallback((input: HTMLInputElement) => {
    if (input !== null) {
      input.focus()
    }
  }, [])

  return (
    <div className="New-member">
      <button type="button" onClick={openOverlay}>
        New user
      </button>
      {opened && (
        <div className="New-member-overlay" onClick={closeOverlay}>
          <div
            className="New-member-content"
            onClick={e => {
              e.stopPropagation()
            }}
          >
            <div className="New-member-content-header">
              <h2>Create member</h2>
              <button type="button" onClick={closeOverlay}>
                X
              </button>
            </div>
            <div className="New-member-content-form">
              <div className="New-member-content-form-input">
                <label htmlFor="firstname">First name</label>
                <input
                  value={newMember.firstname}
                  id="firstname"
                  name="firstname"
                  type="text"
                  onChange={updateMemberPropFromString}
                  tabIndex={1}
                  ref={firstnameInput}
                />
              </div>
              <div className="New-member-content-form-input">
                <label htmlFor="title">Title</label>
                <input
                  value={newMember.title}
                  id="title"
                  name="title"
                  type="text"
                  onChange={updateMemberPropFromString}
                  tabIndex={3}
                />
              </div>
              <div className="New-member-content-form-input">
                <label htmlFor="lastname">Last name</label>
                <input
                  value={newMember.lastname}
                  id="lastname"
                  name="lastname"
                  type="text"
                  onChange={updateMemberPropFromString}
                  tabIndex={2}
                />
              </div>
              <div className="New-member-content-form-input">
                <label htmlFor="business_unit">Business unit</label>
                <input
                  value={newMember.business_unit}
                  id="business_unit"
                  name="business_unit"
                  type="text"
                  onChange={updateMemberPropFromString}
                  tabIndex={4}
                />
              </div>
              <div className="New-member-content-form-input">
                <input
                  checked={newMember.is_user}
                  id="is_user"
                  name="is_user"
                  type="checkbox"
                  onChange={({ target }) => {
                    setNewMember({ ...newMember, is_user: target.checked })
                  }}
                  tabIndex={5}
                />
                <label htmlFor="is_user">User status</label>
                <small>Designates whether the person can login into application</small>
              </div>
            </div>
            <div className="New-member-content-footer">
              <button type="button" onClick={closeOverlay}>
                Cancel
              </button>
              <button type="button" onClick={createNewMember}>
                Create member
              </button>
            </div>
          </div>
        </div>
      )}
    </div>
  )
}
