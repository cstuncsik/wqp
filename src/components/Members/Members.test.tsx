import React from 'react'
import { render, fireEvent } from '@testing-library/react'

import { Members } from 'components/Members/Members'
import { AppContext } from 'AppContext'

describe('Members component', () => {
  it('renders empty list and "Add new member" button disabled when there is no data', () => {
    const { container } = render(<Members />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders empty list and "Add new member" button enabled when there is at least one member', () => {
    const { container } = render(
      <AppContext.Provider
        value={{
          messages: [],
          addMessage: () => {},
          availableMembers: [
            {
              person_id: '1b31592d-3d8a-42ff-9a8a-a373b10c2ade',
              firstname: 'Csaba',
              lastname: 'Tuncsik',
              title: 'Frontend developer',
              business_unit: 'Product',
              is_user: true
            }
          ],
          setAvailableMembers: () => {}
        }}
      >
        <Members />
      </AppContext.Provider>
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders list after clicking "Add new member" button when there is at least one member', () => {
    const { container, getByText } = render(
      <AppContext.Provider
        value={{
          messages: [],
          addMessage: () => {},
          availableMembers: [
            {
              person_id: '1b31592d-3d8a-42ff-9a8a-a373b10c2ade',
              firstname: 'Csaba',
              lastname: 'Tuncsik',
              title: 'Frontend developer',
              business_unit: 'Product',
              is_user: true
            }
          ],
          setAvailableMembers: () => {}
        }}
      >
        <Members />
      </AppContext.Provider>
    )
    fireEvent.click(getByText('Add new member'))
    expect(container.firstChild).toMatchSnapshot()
  })
})
