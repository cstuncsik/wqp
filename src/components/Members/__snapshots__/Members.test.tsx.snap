// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Members component renders empty list and "Add new member" button disabled when there is no data 1`] = `
<div
  class="Members"
>
  <div
    class="Members-data"
  >
    <h2>
      Member
    </h2>
    <h2>
      Role
    </h2>
    <h2>
      Access Level
    </h2>
    <div />
  </div>
  <div
    class="Members-action"
  >
    <button
      disabled=""
      type="button"
    >
      Add new member
    </button>
  </div>
  <div
    class="Members-result"
  >
    <button
      type="button"
    >
      Save result
    </button>
    <pre>
      {
  "members": []
}
    </pre>
  </div>
</div>
`;

exports[`Members component renders empty list and "Add new member" button enabled when there is at least one member 1`] = `
<div
  class="Members"
>
  <div
    class="Members-data"
  >
    <h2>
      Member
    </h2>
    <h2>
      Role
    </h2>
    <h2>
      Access Level
    </h2>
    <div />
  </div>
  <div
    class="Members-action"
  >
    <button
      type="button"
    >
      Add new member
    </button>
  </div>
  <div
    class="Members-result"
  >
    <button
      type="button"
    >
      Save result
    </button>
    <pre>
      {
  "members": []
}
    </pre>
  </div>
</div>
`;

exports[`Members component renders list after clicking "Add new member" button when there is at least one member 1`] = `
<div
  class="Members"
>
  <div
    class="Members-data"
  >
    <h2>
      Member
    </h2>
    <h2>
      Role
    </h2>
    <h2>
      Access Level
    </h2>
    <div />
    <select
      disabled=""
      name="member"
    >
      <option
        value="1b31592d-3d8a-42ff-9a8a-a373b10c2ade"
      >
        Csaba Tuncsik
      </option>
    </select>
    <select
      name="role"
    >
      <option
        value="customer"
      >
        Customer
      </option>
      <option
        value="employee"
      >
        Employee
      </option>
      <option
        value="manager"
      >
        Manager
      </option>
    </select>
    <select
      name="accessLevel"
    >
      <option
        value="read"
      >
        Read
      </option>
      <option
        value="write"
      >
        Write
      </option>
    </select>
    <div>
      <button
        type="button"
      >
        X
      </button>
    </div>
  </div>
  <div
    class="Members-action"
  >
    <button
      disabled=""
      type="button"
    >
      Add new member
    </button>
  </div>
  <div
    class="Members-result"
  >
    <button
      type="button"
    >
      Save result
    </button>
    <pre>
      {
  "members": []
}
    </pre>
  </div>
</div>
`;
